snuba_sdk/expressions.py

```diff
@@ -1,16 +1,52 @@
 from __future__ import annotations
 
 import re
-from dataclasses import dataclass
+from abc import ABC, abstractmethod
+from dataclasses import dataclass, field
 from datetime import date, datetime
-from typing import List, Optional, Set, Union
+from enum import Enum
+from typing import Generic, List, Optional, Set, TypeVar, Union
 
-from snuba_sdk import Expression
 from snuba_sdk.clickhouse import is_aggregation_function
 
 
-class InvalidExpression(Exception):
-    pass
+TVisited = TypeVar("TVisited")
+
+
+class Visitor(ABC, Generic[TVisited]):
```

Comment:
The downside of this is that since the base Expression class wants to validate and translate itself, all of this code has to be in the same file to avoid circular imports. I think future visitors should be able to be in a separate file though.

Reply:
The other alternative is to mimic pattern matching instead of implementing a visitor, just as we do for query classes in snuba: 
https://github.com/getsentry/snuba/blob/master/snuba/query/data_source/visitor.py#L11-L61

This is basically copying over what the visitor does. In language with pattern matching support generally that would give us the same type safety as a visitor (type checking would fail if we add a new type of query, expression here). But there is a way to mimic that in python through mypy exhaustive checks: https://hakibenita.com/python-mypy-exhaustive-checking#type-narrowing-in-mypy
I was not aware of this when I added the data source visitor mimicking pattern matching.

Reply:
I ended up using the pattern matching idea, as well as moving basic validation back into the Expression class. Things are now split into modules, which is good because the Condition and Entity logic will get a lot more complicated in the future. Not a perfect visitor pattern but what are you gonna do.